In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
import csv

#Bảng Segmentation


In [ ]:
Segmentation = pd.read_csv('/content/2017Segmentation3685case.csv',sep=";" )

In [ ]:
Segmentation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4944 entries, 0 to 4943
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ID            4944 non-null   int64 
 1   Segmentation  4944 non-null   object
 2   Visit         4944 non-null   int64 
 3   Spending      4944 non-null   int64 
 4   Brand         4944 non-null   object
 5   PPA           4944 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 231.9+ KB


In [ ]:
Segmentation.head()

ID                           Segmentation  Visit  Spending  \
0   92316  Seg.02 - Mass Asp (VND 25K - VND 59K)      4       120   
1   96307  Seg.02 - Mass Asp (VND 25K - VND 59K)      4       120   
2  105678  Seg.02 - Mass Asp (VND 25K - VND 59K)      4       120   
3  106554  Seg.02 - Mass Asp (VND 25K - VND 59K)      4       120   
4  106555  Seg.02 - Mass Asp (VND 25K - VND 59K)      4       120   

          Brand  PPA  
0  Indepentdent   30  
1  Indepentdent   30  
2  Indepentdent   30  
3  Indepentdent   30  
4  Indepentdent   30

In [ ]:
print(Segmentation.nunique())

ID              3676
Segmentation       4
Visit             49
Spending         225
Brand              3
PPA               56
dtype: int64


In [ ]:
Segmentation.isnull().sum()

ID              0
Segmentation    0
Visit           0
Spending        0
Brand           0
PPA             0
dtype: int64

##In các giá trị unique trong Segmentation để xác định xem có lỗi Logic không

#Bảng Brand_image


In [ ]:
Brand_Image = pd.read_csv('/content/Brand_Image.csv',sep=";" )

In [ ]:
Brand_Image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 643072 entries, 0 to 643071
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   ID          643072 non-null  int64 
 1   Year        643072 non-null  int64 
 2   City        643072 non-null  object
 3   Awareness   642675 non-null  object
 4   Attribute   643072 non-null  object
 5   BrandImage  643072 non-null  object
dtypes: int64(2), object(4)
memory usage: 29.4+ MB


In [ ]:
Brand_Image.head()

ID  Year    City         Awareness      Attribute        BrandImage
0  725118  2019  Hà Nội  Highlands Coffee  Popular brand  Highlands Coffee
1  725466  2019  Hà Nội  Highlands Coffee  Popular brand  Highlands Coffee
2  726561  2019  Hà Nội  Highlands Coffee  Popular brand  Highlands Coffee
3  726862  2019  Hà Nội  Highlands Coffee  Popular brand  Highlands Coffee
4  727219  2019  Hà Nội  Highlands Coffee  Popular brand  Highlands Coffee

In [ ]:
Brand_Image.nunique()

ID            11761
Year              3
City              6
Awareness        37
Attribute        28
BrandImage       37
dtype: int64

In [ ]:
Brand_Image['Awareness'] = Brand_Image['Awareness'].replace(to_replace=r'^Other.*', value='Other', regex=True)
Brand_Image['BrandImage'] = Brand_Image['BrandImage'].replace(to_replace=r'^Other.*', value='Other', regex=True)

In [ ]:
count = Brand_Image['Awareness'] == Brand_Image['BrandImage']

# Số lượng dòng mà hai cột giống nhau
count_equal = count.sum()
print(count_equal)

642675


In [ ]:
Brand_Image.isnull().sum()

ID              0
Year            0
City            0
Awareness     397
Attribute       0
BrandImage      0
dtype: int64

In [ ]:
Brand_Image['Awareness'].value_counts()

Awareness
Other                                            125553
Highlands Coffee                                 106427
Trung Nguyên                                      81577
Street / Half street coffee (including carts)     39747
The Coffee House                                  37529
Cộng Cà Phê                                       35825
Indepedent Cafe                                   33554
Milano                                            30133
Starbucks                                         23643
Phúc Long                                         21931
Street                                            20358
Aha Cafe                                          19114
Urban Station                                     13740
Passio                                             9295
Viva Star                                          7028
Thức Coffee                                        5345
Coffee Bean & Tea Leaf                             3924
Mê Trang                                           3624
Gong Cha                                           3240
Maxx Coffee                                        3123
Long Cafe                                          2511
Mộc Miên                                           2464
Runam cafe                                         1948
Effoc                                              1743
Đen Đá                                             1661
Saigon Café                                        1544
KOI cafe                                           1420
BonPas                                             1381
The Coffee Factory                                 1137
Nia cafe                                            940
Cheese Coffee                                       689
The Cups Coffee                                     388
Laha Coffee                                         139
Name: count, dtype: int64

In [ ]:
Brand_Image.to_csv('Brand_Image_clean.csv', index=False)

Awareness có tống giá trị null là 379, điền bằng cách chuyển từ brandimage sang awareness  
Lý do là vì Awareness là thương hiệu người khảo sát nhận diện, còn BrandImage là người khảo sát nhìn hình để nhận diện  
Do đó nếu khách hàng nhìn hình để dự đoán được thì đã từng nghe được Awareness.

#Bảng BrandHealth


In [ ]:
Brandhealth = pd.read_csv('/content/Brandhealth.csv',sep=";" )

/tmp/ipython-input-10-1590658787.py:1: DtypeWarning: Columns (4,7,8,10,11,12,16,18) have mixed types. Specify dtype option on import or set low_memory=False.
  Brandhealth = pd.read_csv('/content/Brandhealth.csv',sep=";" )


In [ ]:
Brandhealth.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74419 entries, 0 to 74418
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ID                74419 non-null  int64  
 1   Year              74419 non-null  int64  
 2   City              74419 non-null  object 
 3   Brand             74419 non-null  object 
 4   Spontaneous       30993 non-null  object 
 5   Awareness         74305 non-null  object 
 6   Trial             47330 non-null  object 
 7   P3M               28849 non-null  object 
 8   P1M               19399 non-null  object 
 9   Comprehension     26346 non-null  object 
 10  Brand_Likability  10331 non-null  object 
 11  Weekly            13382 non-null  object 
 12  Daily             7621 non-null   object 
 13  Fre#visit         19332 non-null  float64
 14  PPA               14073 non-null  float64
 15  Spending          14073 non-null  float64
 16  Segmentation      14073 non-null  object

In [ ]:
Brandhealth['Brand'].value_counts()

Brand
Trung Nguyên                                     9954
Highlands Coffee                                 8520
Street / Half street coffee (including carts)    7391
Other 1                                          5536
Milano                                           4699
Cộng Cà Phê                                      3929
The Coffee House                                 3830
Starbucks                                        3350
Street                                           3278
Other 2                                          2861
Phúc Long                                        2786
Indepedent Cafe                                  2717
Aha Cafe                                         1852
Urban Station                                    1667
Passio                                           1627
Thức Coffee                                      1061
Viva Star                                        1029
Other 3                                           840
Mê Trang                                          836
Coffee Bean & Tea Leaf                            824
Long Cafe                                         710
Gong Cha                                          594
Mộc Miên                                          585
Đen Đá                                            520
Other Branded Cafe Chain                          496
Effoc                                             473
Maxx Coffee                                       397
Saigon Café                                       363
Runam cafe                                        316
BonPas                                            316
Nia cafe                                          263
The Coffee Factory                                256
KOI cafe                                          237
The Cups Coffee                                   133
Cheese Coffee                                     123
Laha Coffee                                        50
Name: count, dtype: int64

In [ ]:
# Lọc các dòng có Brand là "Highland"
highland_rows = Brandhealth[Brandhealth['Brand'] == 'Highlands Coffee']

# In kết quả
print(highland_rows)
highland_rows = highland_rows.dropna(subset=['Segmentation'])
highland_rows.to_csv('highland_rows.csv', index=False)

           ID  Year         City             Brand       Spontaneous  \
1149   345938  2018  Hồ Chí Minh  Highlands Coffee               NaN   
1150   349764  2018  Hồ Chí Minh  Highlands Coffee               NaN   
1151   349957  2018  Hồ Chí Minh  Highlands Coffee               NaN   
1152   350090  2018  Hồ Chí Minh  Highlands Coffee               NaN   
1153   350576  2018  Hồ Chí Minh  Highlands Coffee               NaN   
...       ...   ...          ...               ...               ...   
74212  458059  2018       Hà Nội  Highlands Coffee  Highlands Coffee   
74213  459806  2018  Hồ Chí Minh  Highlands Coffee  Highlands Coffee   
74246  400654  2018    Hải Phòng  Highlands Coffee  Highlands Coffee   
74286  447981  2018    Nha Trang  Highlands Coffee  Highlands Coffee   
74293  450932  2018      Đà Nẵng  Highlands Coffee  Highlands Coffee   

              Awareness             Trial               P3M               P1M  \
1149   Highlands Coffee               NaN             

In [ ]:
Brandhealth.isnull().sum()

ID                      0
Year                    0
City                    0
Brand                   0
Spontaneous         43426
Awareness             114
Trial               27089
P3M                 45570
P1M                 55020
Comprehension       48073
Brand_Likability    64088
Weekly              61037
Daily               66798
Fre#visit           55087
PPA                 60346
Spending            60346
Segmentation        60346
NPS#P3M             52814
NPS#P3M#Group       52814
Spending_use        60346
dtype: int64

Điền Awareness bằng giá trị thường gặp  
Còn lại thay thế bằng "Không"

In [ ]:
ans1= Brandhealth['Awareness'] == Brandhealth['Brand']
count_ans = ans1.sum()
print(count_ans)

70562


Other 1 2 3.. chuyển thành other  
Chuyển từ brand sang awareness  
Brand là chính  
P3M, P1M: Nếu có điền là số 1, không điền là số 0  


#Bảng Companion

In [ ]:
Companion = pd.read_csv('/content/Companion.csv',sep=";" )

In [ ]:
Companion.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20739 entries, 0 to 20738
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   ID               20739 non-null  int64 
 1   City             20739 non-null  object
 2   Companion#group  20739 non-null  object
 3   Year             20739 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 648.2+ KB


In [ ]:
Companion.nunique()

ID                 11746
City                   6
Companion#group        6
Year                   3
dtype: int64

In [ ]:
Companion.isnull().sum()

ID                 0
City               0
Companion#group    0
Year               0
dtype: int64

#Bảng Competitor

In [ ]:
Competitor = pd.read_csv('/content/Competitor database_xlnm#_FilterDatabase.csv',sep=";" )

In [ ]:
Competitor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234 entries, 0 to 233
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   No#         234 non-null    int64 
 1   Brand       234 non-null    object
 2   City        234 non-null    object
 3   Year        234 non-null    int64 
 4   StoreCount  234 non-null    int64 
dtypes: int64(3), object(2)
memory usage: 9.3+ KB


In [ ]:
Competitor.isnull().sum()

No#           0
Brand         0
City          0
Year          0
StoreCount    0
dtype: int64

In [ ]:
# Tính tổng StoreCount cho mỗi Brand
brand_store_total = Competitor.groupby('Brand')['StoreCount'].sum().reset_index()

# Sắp xếp theo tổng số lượng store giảm dần (nếu muốn)
brand_store_total = brand_store_total.sort_values(by='StoreCount', ascending=False)

# Hiển thị kết quả
print(brand_store_total)


              Brand  StoreCount
1  Highlands Coffee         797
4  The Coffee House         354
5      Trung Nguyên         177
0       Cộng Cà Phê         162
3         Starbucks         132
2         Phúc Long         118


#Bảng Dayofweek

In [ ]:
Dayofweek = pd.read_csv('/content/Dayofweek.csv',sep=";" )

In [ ]:
Dayofweek.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39095 entries, 0 to 39094
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               39095 non-null  int64  
 1   City             39095 non-null  object 
 2   Dayofweek        39009 non-null  object 
 3   Visit#Dayofweek  39041 non-null  float64
 4   Year             39095 non-null  int64  
 5   Weekday#end      39009 non-null  object 
dtypes: float64(1), int64(2), object(3)
memory usage: 1.8+ MB


In [ ]:
Dayofweek.nunique()

ID                 11761
City                   6
Dayofweek              7
Visit#Dayofweek       26
Year                   3
Weekday#end            2
dtype: int64

In [ ]:
Dayofweek.isnull().sum()

ID                  0
City                0
Dayofweek          86
Visit#Dayofweek    54
Year                0
Weekday#end        86
dtype: int64

In [ ]:
Dayofweek['Dayofweek'].value_counts()

Dayofweek
Sunday       9246
Saturday     8630
Thursday     4449
Friday       4429
Wednesday    4240
Monday       4075
Tuesday      3940
Name: count, dtype: int64

In [ ]:
Dayofweek['Weekday#end'].value_counts()

Weekday#end
Weekdays    21133
Weekends    17876
Name: count, dtype: int64

Đầu tiên điền Weekday#end trước, bằng phương pháp

#Bảng Daypart

In [ ]:
Daypart = pd.read_csv('/content/Daypart.csv',sep=";" )

In [ ]:
Daypart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19189 entries, 0 to 19188
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ID             19189 non-null  int64  
 1   City           19189 non-null  object 
 2   Daypart        19176 non-null  object 
 3   Visit#Daypart  18342 non-null  float64
 4   Year           19189 non-null  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 749.7+ KB


In [ ]:
Daypart.nunique()

ID               11761
City                 6
Daypart              6
Visit#Daypart       39
Year                 3
dtype: int64

In [ ]:
Daypart.isnull().sum()

ID                 0
City               0
Daypart           13
Visit#Daypart    847
Year               0
dtype: int64

#Bảng NeedstateDayDaypart

In [ ]:
Need = pd.read_csv('/content/NeedstateDayDaypart.csv',sep=";" )

In [ ]:
Need.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75251 entries, 0 to 75250
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ID              75251 non-null  int64 
 1   City            75251 non-null  object
 2   Year            75251 non-null  int64 
 3   Needstates      75251 non-null  object
 4   Day#Daypart     75251 non-null  object
 5   NeedstateGroup  75251 non-null  object
dtypes: int64(2), object(4)
memory usage: 3.4+ MB


In [ ]:
Need.nunique()

ID                11740
City                  6
Year                  3
Needstates           15
Day#Daypart           9
NeedstateGroup        6
dtype: int64

In [ ]:
Need.isnull().sum()

ID                0
City              0
Year              0
Needstates        0
Day#Daypart       0
NeedstateGroup    0
dtype: int64

#Bảng SA

In [ ]:
sa = pd.read_csv('/content/SA#var.csv',sep=";" )

In [ ]:
sa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11761 entries, 0 to 11760
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ID                11761 non-null  int64  
 1   City              11761 non-null  object 
 2   Group_size        11746 non-null  float64
 3   Age               11752 non-null  float64
 4   MPI#Mean          8044 non-null   float64
 5   TOM               11761 non-null  object 
 6   BUMO              11761 non-null  object 
 7   BUMO_Previous     6096 non-null   object 
 8   MostFavourite     11761 non-null  object 
 9   Gender            11761 non-null  object 
 10  MPI#detail        8076 non-null   object 
 11  Age#group         11752 non-null  object 
 12  Age#Group#2       11752 non-null  object 
 13  MPI               8044 non-null   object 
 14  MPI#2             8044 non-null   object 
 15  Occupation        11761 non-null  object 
 16  Occupation#group  11761 non-null  object

In [ ]:
sa.nunique()

ID                  11761
City                    6
Group_size             16
Age                    45
MPI#Mean               13
TOM                    32
BUMO                   33
BUMO_Previous          32
MostFavourite          34
Gender                  2
MPI#detail             14
Age#group               4
Age#Group#2             7
MPI                     5
MPI#2                   5
Occupation             28
Occupation#group        7
Year                    3
Col                     6
MPI_Mean_Use           13
dtype: int64

In [ ]:
sa.isnull().sum()

ID                     0
City                   0
Group_size            15
Age                    9
MPI#Mean            3717
TOM                    0
BUMO                   0
BUMO_Previous       5665
MostFavourite          0
Gender                 0
MPI#detail          3685
Age#group              9
Age#Group#2            9
MPI                 3717
MPI#2               3717
Occupation             0
Occupation#group       0
Year                   0
Col                    0
MPI_Mean_Use        3717
dtype: int64

Age, Age#group, Age#group2 không đụng
Kiểm tra MPI với MPI detail, nếu như MPI có mà MPI detail không có => điền  


##Xử lý các giá trị null

Bảng segmentation: Không có giá trị null, không có giá trị vi phạm lỗi logic

Bảng Brand_image: Thuộc tính Awareness có 44 giá trị null, xử lý bằng cách gán giá trị từ Brandimage qua


In [ ]:
for index, row in Brand_Image.iterrows():
    if pd.isnull(row['Awareness']):
        Brand_Image.at[index, 'Awareness'] = row['BrandImage']
print(Brand_Image.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 643072 entries, 0 to 643071
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   ID          643072 non-null  int64 
 1   Year        643072 non-null  int64 
 2   City        643072 non-null  object
 3   Awareness   643072 non-null  object
 4   Attribute   643072 non-null  object
 5   BrandImage  643072 non-null  object
dtypes: int64(2), object(4)
memory usage: 29.4+ MB
None


Bảng Brandhealth: Chưa xử lý 😁

Bảng Companion: Không có giá trị khuyết

Bảng Competitor: Không có giá trị khuyết

Bảng Dayofweek: Thuộc tính "Dayofweek" có 86 giá trị khuyết, Thuộc tính "Visit#Dayofweek" có 54 giá trị khuyết, thuộc tính "Weekday#end" có 86 giá trị khuyết  
Chúng ta sẽ điền giá trị "Dayofweek", sau đó suy ra được "Weekday"  
Dayofweek sử dụng mô hình hồi quy   
Visit#Dayofweek sử dụng mô hình hồi quy

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

# Giả sử bạn đã có DataFrame df
df = pd.read_csv('/content/Dayofweek.csv', sep=';')

# -------------------------------
# BƯỚC 1: Điền khuyết Dayofweek và Weekday#end nếu chỉ thiếu 1 trong 2
# -------------------------------

weekday_days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']
weekend_days = ['Saturday', 'Sunday']

# Map Weekday#end từ Dayofweek
weekday_map = {
    'Monday': 'Weekday',
    'Tuesday': 'Weekday',
    'Wednesday': 'Weekday',
    'Thursday': 'Weekday',
    'Friday': 'Weekday',
    'Saturday': 'Weekend',
    'Sunday': 'Weekend'
}

# a. Nếu Weekday#end bị thiếu, dùng Dayofweek để điền
mask_weekend_missing = df['Weekday#end'].isna() & df['Dayofweek'].notna()
df.loc[mask_weekend_missing, 'Weekday#end'] = df.loc[mask_weekend_missing, 'Dayofweek'].map(weekday_map)

# b. Nếu Dayofweek bị thiếu, dùng Weekday#end để điền
mask_day_missing = df['Dayofweek'].isna() & df['Weekday#end'].notna()
df.loc[mask_day_missing, 'Dayofweek'] = df.loc[mask_day_missing, 'Weekday#end'].apply(
    lambda x: np.random.choice(weekday_days) if x == 'Weekday' else np.random.choice(weekend_days)
)

# -------------------------------
# BƯỚC 2: Dự đoán khi cả 2 cột bị thiếu (mô hình logic)
# -------------------------------

# Chọn tập train: không bị thiếu
train_df = df[df['Weekday#end'].notna() & df['Visit#Dayofweek'].notna()]

X_train = train_df[['City', 'Visit#Dayofweek', 'Year']]
y_train = train_df['Weekday#end']

cat_cols = ['City']
num_cols = ['Visit#Dayofweek', 'Year']

preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), cat_cols),
        ('num', 'passthrough', num_cols)
    ]
)

clf_weekend = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(max_iter=1000))
])

clf_weekend.fit(X_train, y_train)

# Dự đoán các dòng thiếu cả 2 cột
missing_both = df[df['Weekday#end'].isna() & df['Dayofweek'].isna()]
X_missing = missing_both[['City', 'Visit#Dayofweek', 'Year']]
X_missing_valid = X_missing.dropna()
valid_idx = X_missing_valid.index

predicted_weekend = clf_weekend.predict(X_missing_valid)
df.loc[valid_idx, 'Weekday#end'] = predicted_weekend
df.loc[valid_idx, 'Dayofweek'] = df.loc[valid_idx, 'Weekday#end'].apply(
    lambda x: np.random.choice(weekday_days) if x == 'Weekday' else np.random.choice(weekend_days)
)

# -------------------------------
# BƯỚC 3: Điền khuyết cột Visit#Dayofweek bằng mô hình hồi quy
# -------------------------------

# Huấn luyện model hồi quy
train_reg = df[df['Visit#Dayofweek'].notna() & df['Dayofweek'].notna() & df['Weekday#end'].notna()]
X_reg = train_reg[['City', 'Year', 'Dayofweek', 'Weekday#end']]
y_reg = train_reg['Visit#Dayofweek']

cat_cols_reg = ['City', 'Dayofweek', 'Weekday#end']
num_cols_reg = ['Year']

preprocessor_reg = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), cat_cols_reg),
        ('num', 'passthrough', num_cols_reg)
    ]
)

reg_model = Pipeline(steps=[
    ('preprocessor', preprocessor_reg),
    ('regressor', LinearRegression())
])

reg_model.fit(X_reg, y_reg)

# Dự đoán các dòng thiếu Visit#Dayofweek
missing_visit = df[df['Visit#Dayofweek'].isna() & df['City'].notna() & df['Dayofweek'].notna() & df['Weekday#end'].notna()]
X_missing_reg = missing_visit[['City', 'Year', 'Dayofweek', 'Weekday#end']]
predicted_visits = reg_model.predict(X_missing_reg)

# Gán giá trị dự đoán và làm tròn về số nguyên >= 0
df.loc[X_missing_reg.index, 'Visit#Dayofweek'] = np.maximum(0, np.round(predicted_visits))

# -------------------------------
# BƯỚC 4: Kết quả cuối cùng
# -------------------------------
print("Số giá trị thiếu còn lại sau khi xử lý:")
print(df[['Dayofweek', 'Weekday#end', 'Visit#Dayofweek']].isna().sum())

Số giá trị thiếu còn lại sau khi xử lý:
Dayofweek          0
Weekday#end        0
Visit#Dayofweek    0
dtype: int64


In [ ]:
df.to_csv('Dayofweek_clean.csv', index=False)

In [ ]:
for index, row in df.iterrows():
    if row['Dayofweek'] in ['Saturday', 'Sunday'] and row['Weekday#end'] == 'Weekdays':
        df.at[index, 'Weekday#end'] = 'Weekends'

In [ ]:
print(df['Dayofweek'].value_counts())
df['Weekday#end'].value_counts()

Dayofweek
Sunday       9284
Saturday     8678
Thursday     4449
Friday       4429
Wednesday    4240
Monday       4075
Tuesday      3940
Name: count, dtype: int64


Weekday#end
Weekdays    21133
Weekends    17962
Name: count, dtype: int64

In [ ]:
df.to_csv('ten_file2.csv', index=False)

Bảng Daypart: Thuộc tính "Daypart" có 13 giá trị null, thuộc tính "Visit#Daypart" có 847 giá trị null  
Daypart sử dụng pp điển bằng giá trị mode  
Visit#Daypart sử dụng pp median vì đây là dữ liệu mang tính chất tần suất

In [ ]:
Daypart['Visit#Daypart'].value_counts()

Visit#Daypart
2.0     3636
4.0     3110
3.0     2223
1.0     1924
5.0     1319
10.0     994
6.0      826
30.0     800
8.0      705
7.0      409
20.0     375
15.0     336
24.0     323
12.0     287
25.0     218
9.0      133
26.0     110
16.0      87
11.0      75
28.0      74
14.0      64
18.0      61
13.0      56
22.0      55
17.0      34
21.0      28
19.0      17
23.0      16
27.0      12
40.0      12
60.0       8
29.0       5
35.0       3
31.0       2
48.0       1
34.0       1
45.0       1
38.0       1
43.0       1
Name: count, dtype: int64

In [ ]:
DPmedian = Daypart['Visit#Daypart'].median()
Daypart['Visit#Daypart'].fillna(DPmedian, inplace = True)

/tmp/ipython-input-39-1419924755.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  Daypart['Visit#Daypart'].fillna(DPmedian, inplace = True)


In [ ]:
Daypart.isnull().sum()

ID               0
City             0
Daypart          0
Visit#Daypart    0
Year             0
dtype: int64

In [ ]:
Daypart['Daypart'].fillna('5 PM - before 9 PM', inplace = True)

/tmp/ipython-input-41-4284341440.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  Daypart['Daypart'].fillna('5 PM - before 9 PM', inplace = True)


In [ ]:
Daypart['Daypart'].value_counts()

Daypart
5 PM - before 9 PM     6604
Before 9 AM            5698
9 AM - before 11 AM    2370
2 PM - before 5 PM     2042
11 AM - before 2 PM    1644
9 PM or later           831
Name: count, dtype: int64

In [ ]:
Daypart.to_csv('Daypart_clean.csv', index=False)

Bảng NeedstateDayDaypart: Không có giá trị khuyết


Bảng SA: nhiều quá chưa xử lý 😅😅